### PR TITLE
Adopt Zabbix protocol

### LINKED ISSUE #6118: Adopt Zabbix protocol
Zabbix is a traditional and widely used metrics monitoring mechanism. We should consider to adopt its network protocol, and convert to our metrics system. 
All @apache/skywalking-committers , this requirement will need more discussion on tech details. If any of you feels interested in this, please reply.

NOTE, Zabbix is GPL licensed, but because we don't copy any codes, just implement its protocol in our own codes, and optional feature only. it is not an issue.

### PR DESCRIPTION
**Pull request by mrproliu**
### Adopt Zabbix protocol
- [x] Update the documentation to include this new feature.
- [x] Tests(including UT, IT, E2E) are added to verify the new feature.
- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #6118 
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
I removed the 8.4.0 for now, as we don't have the VM concept for now. FYI @wankai123 , this depends on your side process.

**Comment by wu-sheng** [comment]
https://twitter.com/ASFSkyWalking/status/1355129922356830210

This should be a good example for you to match Zabbix meitrcs with @wankai123 's OTEL metrics for the VM. I hope those could be unified.

**Comment by mrproliu** [comment]
> https://twitter.com/ASFSkyWalking/status/1355129922356830210
> 
> This should be a good example for you to match Zabbix meitrcs with @wankai123 's OTEL metrics for the VM. I hope those could be unified.

Sure. I will follow the metrics, and try to keep Zabbix metrics consistent.

File: docs/en/setup/backend/backend-zabbix.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+## Module define
+```yaml
+receiver-zabbix:
+  selector: ${SW_RECEIVER_METER:default}
+  default:
+    # Export tcp port, Zabbix agent could connected and transport data
+    port: 13800
+    # Enable config when receive agent request
+    activeFiles: agent.yaml
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-zabbix.md:12]: Please take a reference of OTEL rules, I remember it uses file name without `.yaml`.

File: docs/en/setup/backend/backend-zabbix.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+```
+
+## Configuration file
+Zabbix receiver is configured via a configuration file. The configuration file defines everything related to receiving 
+ from agents, as well as which rule files to load.
+ 
+OAP can load the configuration at bootstrap. If the new configuration is not well-formed, OAP fails to start up. The files
+are located at `$CLASSPATH/zabbix-receive-config`.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-zabbix.md:20]:
```suggestion
are located at `$CLASSPATH/zabbix-rules`.
```

File: docs/en/setup/backend/backend-zabbix.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+The file is written in YAML format, defined by the scheme described below. Brackets indicate that a parameter is optional.
+
+A example can be found [here](../../../../oap-server/server-bootstrap/src/main/resources/meter-analyzer-config/spring-sleuth.yaml).
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-zabbix.md:24]: Copy/paste issue? I think this is not related to sleuth.

Comment (wu-sheng) [reply docs/en/setup/backend/backend-zabbix.md:24]: We could provide native configuration for some Zabbix agents, such as VM agent

File: docs/en/setup/backend/backend-zabbix.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+You could find the Zabbix agent detail items from [Zabbix Agent docucment](https://www.zabbix.com/documentation/current/manual/config/items/itemtypes/zabbix_agent).
+
+### Zabbix agent configure
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-zabbix.md:27]: This title could mislead people thought, this is Zabbix's native configuration. This should be purely `Configuration file`

File: oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/protocol/ZabbixProtocolEncoder.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            (byte) (payloadLength >> 8 & 0xFF),
+            (byte) (payloadLength >> 16 & 0xFF),
+            (byte) (payloadLength >> 24 & 0xFF),
+            '\0', '\0', '\0', '\0'};
+
+        // Build and write ByteBuf
+        ByteBuf buffer = channelHandlerContext.alloc().buffer(header.length + payloadLength);
+        buffer.writeBytes(header);
+        buffer.writeBytes(responsePayload.getBytes(Charsets.UTF_8));
+
```
Comment (kezhenxu94) [inline oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/protocol/ZabbixProtocolEncoder.java:54]:
```suggestion
        buffer.retain();
```

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java [hunk 6]
```diff
@@ -105,12 +106,11 @@ public SampleFamily propertyMissing(String metricName) {
...
                 if (sampleFamilies.containsKey(metricName)) {
+                    ExpressionParsingContext.get().ifPresent(ctx -> ctx.samples.add(metricName));
                     return sampleFamilies.get(metricName);
                 }
-                if (!ExpressionParsingContext.get().isPresent()) {
-                    log.warn("{} referred by \"{}\" doesn't exist in {}", metricName, literal, sampleFamilies.keySet());
-                }
-                return SampleFamily.EMPTY;
+                // Using finder, support "."
+                return new SampleFamilyFinder(metricName, literal, sampleFamilies);
```
Comment (wu-sheng) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java:113]: Do you mean, `is there is no method found than failing back to a sample name including dot?`?

Comment (wu-sheng) [reply oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java:113]: If so, you are facing conflict with Prometheus and Telemetry rules.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# UI templates initialized file includes the default template when the SkyWalking OAP starts up at the first time.
+#
+# Also, SkyWalking would detect the existing templates in the database, once they are missing, all templates in this file
+# could be added automatically.
+
+templates:
+  - name: "Zabbix"
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml:22]: Do you really need this? My expectation would be, the VM monitoring could share the same templates with existing VM template. Could you share the details about why adding this?

File: docs/en/setup/backend/backend-zabbix.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
+# Zabbix Receiver
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-zabbix.md:1]: Do we have some benchmark about this? And which authentication and security are we supporting now?

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamilyFinder.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+import groovy.lang.GroovyInterceptable;
+import groovy.lang.MetaClass;
+import lombok.extern.slf4j.Slf4j;
+import org.codehaus.groovy.runtime.InvokerHelper;
+
+/**
+ * Help to find {@link SampleFamily}, support "." in metric name
+ */
+@Slf4j
+public class SampleFamilyFinder extends SampleFamily implements GroovyInterceptable {
```
Comment (wu-sheng) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamilyFinder.java:31]: I have concern of this mechanism, let's discuss this first.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java [hunk 10]
```diff
@@ -284,6 +289,11 @@ public void prepare() throws ServiceNotProvidedException, ModuleStartException {
         TopNStreamProcessor.getInstance().setTopNWorkerReportCycle(moduleConfig.getTopNReportPeriod());
         apdexThresholdConfig = new ApdexThresholdConfig(this);
         ApdexMetrics.setDICT(apdexThresholdConfig);
+
+        // TCP manager
+        tcpServerManager = new TCPServerManager(moduleConfig.getTcpHost(),
+            moduleConfig.getTcpBossGroupThreadCount(), moduleConfig.getTcpWorkerGroupThreadCount());
+        this.registerServiceImplementation(TCPBinderRegister.class, new TCPBinderRegisterImpl(tcpServerManager));
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java:296]: Why TCP server is being injected in the core level? TCP(a.k.a. private network protocol) server for a specific ecosystem only.

File: docs/en/setup/backend/backend-zabbix.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+Zabbix receiver is accepting the metrics of [Zabbix Agent Active Checks protocol](https://www.zabbix.com/documentation/current/manual/appendix/items/activepassive#active_checks) fromat into the [Meter System](./../../concepts-and-designs/meter.md).
+
+## Module define
+```yaml
+receiver-zabbix:
+  selector: ${SW_RECEIVER_METER:default}
+  default:
+    # Export tcp port, Zabbix agent could connected and transport data
+    port: 13800
```
[the remark this answers is not in the document]
Comment (mrproliu) [reply docs/en/setup/backend/backend-zabbix.md:10]: The port is `10051`, should we use this port?

File: test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/kafka/KafkaE2E.java [hunk 12]
```diff
@@ -200,8 +202,9 @@ void so11y() throws Exception {
                 for (String metricsName : ALL_SO11Y_LABELED_METRICS) {
                     LOGGER.info("verifying service instance response time: {}", instance);
                     final List<ReadMetrics> instanceMetrics = graphql.readLabeledMetrics(
-                        new ReadMetricsQuery().stepByMinute().metricsName(metricsName)
+                        new ReadLabeledMetricsQuery().stepByMinute().metricsName(metricsName)
                                               .serviceName(service.getLabel()).instanceName(instance.getLabel())
+                                              .labels(Arrays.asList("50", "70", "90", "99"))
```
Comment (wu-sheng) [inline test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/kafka/KafkaE2E.java:207]: What is this? A previous bug?

File: test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/storage/StorageE2E.java [hunk 13]
```diff
@@ -262,8 +264,9 @@ void so11y() throws Exception {
                 for (String metricsName : ALL_SO11Y_LABELED_METRICS) {
                     LOGGER.info("verifying service instance response time: {}", instance);
                     final List<ReadMetrics> instanceMetrics = graphql.readLabeledMetrics(
-                        new ReadMetricsQuery().stepByMinute().metricsName(metricsName)
+                        new ReadLabeledMetricsQuery().stepByMinute().metricsName(metricsName)
                             .serviceName(service.getLabel()).instanceName(instance.getLabel())
+                            .labels(Arrays.asList("50", "70", "90", "99"))
```
Comment (wu-sheng) [inline test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/storage/StorageE2E.java:269]: Same question here.

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java (hunk 6 above)
Comment (mrproliu) [reply oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java:113]: If there no property(not the method) found, it will append `"." + propertyName` to the meter name and trying to find.
If the method not found, it will same as the `SampleFamily`.

File: test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/kafka/KafkaE2E.java (hunk 12 above)
Comment (mrproliu) [reply test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/kafka/KafkaE2E.java:207]: The previous version of the labeled metrics query is not supported `label` customize. So I just add a new query type and add the `labels` property to customize in the `gql` file.

File: docs/en/setup/backend/backend-zabbix.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (wu-sheng) [reply docs/en/setup/backend/backend-zabbix.md:10]: Yes, we are providing an alternative, so, same port would be better.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml (hunk 7 above)
Comment (mrproliu) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml:22]: There has a little difference between OTEL or Zabbix. Such as `Disk R/W`, the Zabbix is not supporting read bytes per second, only support read/write count. I found the [Zabbix agent support platform document](https://www.zabbix.com/documentation/current/manual/appendix/items/supported_by_platform), It could behave more description.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java (hunk 10 above)
Comment (mrproliu) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java:296]: Here is inject the TCP server manager, all TCP server depends on `TCPBinder`. Declare in here because I need to share the TCP connect and handle thread pool.

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java:296]: > because I need to share the TCP connect

I don't agree, what case you could share 2 kinds of TCP protocol with one server?

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml (hunk 7 above)
Comment (wu-sheng) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml:22]: image

I can see the supports of read and write, what do I miss?

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java (hunk 6 above)
Comment (wu-sheng) [reply oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java:113]: Take a look how Prometheus metrics family did. All `.`s are replaced by `_` before the MAL engine.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java (hunk 10 above)
Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java:296]: Zabbix server with its TCP protocol is very unique, how this could share with others? Could you provide a real or closing real(in theory) case to support this?

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml (hunk 7 above)
Comment (mrproliu) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml:22]: That's true, but the `type` parameter value `bytes` Linux is not supported.
image
image

Comment (wu-sheng) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/zabbix.yml:22]: Let's merge sps into the existing the disk metrics. We could declare them clearly on the UI configuration, rather than providing a totally new dashboard.

**Comment by mrproliu** [comment]
I have tested, and take a screenshot in the VM view.
image

**Comment by wu-sheng** [comment]
It looks nice. Have you tested hybrid using Prometheus Node exporter and Zabbix? Does MAL support this(Same metrics in different MAL files)? @hanahmily

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java (hunk 6 above)
Comment (mrproliu) [reply oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/Expression.java]: Replace all `.` to `_` in the Zabbix metric name.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java (hunk 10 above)
Comment (mrproliu) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java]: Change to define TCP protocol in the Zabbix receive module.

File: docs/en/setup/backend/backend-zabbix.md (hunk 8 above)
Comment (mrproliu) [reply docs/en/setup/backend/backend-zabbix.md:1]: It using TCP short connection. no plan to add a benchmark. And no authentication support currently.

File: docs/en/setup/backend/backend-zabbix.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+```yaml
+receiver-zabbix:
+  selector: ${SW_RECEIVER_ZABBIX:default}
+  default:
+    # Export tcp port, Zabbix agent could connected and transport data
+    port: 10051
+    # Bind to host
+    host: 0.0.0.0
+    # Enable config when receive agent request
+    activeFiles: agent
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-zabbix.md:14]: All configuration should be in `configuration-vocalbulary` doc.

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java [hunk 15]
```diff
@@ -60,7 +60,7 @@
 /**
  * SampleFamily represents a collection of {@link Sample}.
  */
-@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
```
Comment (wu-sheng) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java:63]: Why change this?

File: oap-server/server-bootstrap/src/main/resources/zabbix-rules/agent.yaml [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    exp: system_swap_size.tagEqual('2', 'total').avg(['host'])
+  - name: memory_swap_percentage
+    exp: system_swap_size.tagEqual('2', 'pused')
+  # file
+  - name: filesystem_percentage
+    exp: vfs_fs_inode.tagEqual('2', 'pused').avg(['1', 'host'])
+  - name: vfs_fs_size
+    exp: vfs_fs_size.avg(['1', '2', 'host'])
+  - name: disk_read
+    # Adapt the VM UI template
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/zabbix-rules/agent.yaml:85]:
```suggestion
    # `* 1024` is for adapting to `divide 1024` by the VM UI template configuration, which converts `byte`->`kb` because of OTEL/Prometheus Node exporter.
```

File: oap-server/server-bootstrap/src/main/resources/zabbix-rules/agent.yaml [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+    exp: vfs_dev_read * 1024
+  - name: disk_written
+    # Adapt the VM UI template
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/zabbix-rules/agent.yaml:88]: Same here.

File: oap-server/server-bootstrap/src/main/resources/application.yml [hunk 18]
```diff
@@ -266,6 +266,13 @@ receiver-profile:
   selector: ${SW_RECEIVER_PROFILE:default}
   default:
 
+receiver-zabbix:
+  selector: ${SW_RECEIVER_ZABBIX:default}
+  default:
+    port: ${SW_RECEIVER_ZABBIX_PORT:10051}
+    host: ${SW_RECEIVER_ZABBIX_HOST:0.0.0.0}
+    activeFiles: agent
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/application.yml:274]: Why this doesn't include system_env format?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModule.java [hunk 19]
```diff
@@ -55,6 +53,9 @@
 import org.apache.skywalking.oap.server.core.worker.IWorkerInstanceSetter;
 import org.apache.skywalking.oap.server.library.module.ModuleDefine;
 
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModule.java:58]: You are not changing this.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleConfig.java [hunk 20]
```diff
@@ -134,6 +135,19 @@
...
+    private String tcpHost = "0.0.0.0";
+    /**
+     * The count of TCP boss group thread to building connection channel.
+     */
+    private int tcpBossGroupThreadCount = 1;
+    /**
+     * The count of TCP worker group thread to handle data.
+     */
+    private int tcpWorkerGroupThreadCount = 0;
+
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleConfig.java:150]: This should not exist here.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java [hunk 21]
```diff
@@ -98,6 +94,11 @@
 import org.apache.skywalking.oap.server.telemetry.TelemetryModule;
 import org.apache.skywalking.oap.server.telemetry.api.TelemetryRelatedContext;
 
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java:101]: You should not change this.

File: oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/ZabbixMetrics.java [hunk 22]
```diff
@@ -0,0 +1,319 @@
...
+
+    /**
+     * All enabled service and instance group
+     */
+    private List<InstanceGroup> allServices = new ArrayList<>();
+
+    /**
+     * Cache host name to service group, help to quick find service group when receive agent request
+     */
+    private Map<String, InstanceGroup> hostWithGroupCache = new ConcurrentHashMap<>();
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/ZabbixMetrics.java:76]: You wouldn't have much rules to use. It is better to reduce the complexity and remove these caches. I think Regex is fast enough in this case.

**Comment by wu-sheng** [comment]
> It looks nice. Have you tested hybrid using Prometheus Node exporter and Zabbix? Does MAL support this(Same metrics in different MAL files)? @hanahmily

Checked, no supported yet, add https://github.com/apache/skywalking/issues/6356 to enhance this.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/vm.yml [hunk 23]
```diff
@@ -154,7 +154,7 @@ templates:
                   "chartType": "ChartLine",
                   "aggregation": "/",
                   "aggregationNum": "1024",
-                  "unit": "KB/s"
+                  "unit": "KB/s - OTEL, ops - Zabbix"
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates/vm.yml:157]: FYI @wankai123

File: oap-server/server-bootstrap/src/main/resources/zabbix-rules/agent.yaml [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+  - name: filesystem_percentage
+    exp: vfs_fs_inode.tagEqual('2', 'pused').avg(['1', 'host'])
+  - name: vfs_fs_size
+    exp: vfs_fs_size.avg(['1', '2', 'host'])
+  - name: disk_read
+    # `* 1024` is for adapting to `divide 1024` by the VM UI template configuration, which converts `byte`->`kb` because of OTEL/Prometheus Node exporter.
+    exp: vfs_dev_read * 1024
+  - name: disk_written
+    # `* 1024` is for adapting to `divide 1024` by the VM UI template configuration, which converts `byte`->`kb` because of OTEL/Prometheus Node exporter.
+    exp: vfs_dev_write * 1024
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/zabbix-rules/agent.yaml:89]: FYI @wankai123

File: oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/protocol/ZabbixProtocolDecoder.java [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+
+    /**
+     * Close connection if protocol error
+     */
+    private void errorProtocol(ChannelHandlerContext context, ByteBuf byteBuf, String reason) throws InterruptedException {
+        log.warn("Receive message is not Zabbix protocol, reason: {}", reason);
+        // Skip all content
+        byteBuf.skipBytes(byteBuf.readableBytes());
+        // Close connection
+        context.close();
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/protocol/ZabbixProtocolDecoder.java:116]: Don't we need `context.close()` any other place?  Current code structure seems having connection leak risk. Could you recheck?

File: oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/ZabbixMetrics.java [hunk 26]
```diff
@@ -0,0 +1,312 @@
...
+            this.total += statics.total;
+            this.success += statics.success;
+            this.failed += statics.failed;
+            this.useTime += statics.useTime;
+            return this;
+        }
+    }
+
+    /**
+     * Instance group metrics
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/ZabbixMetrics.java:145]:
```suggestion
     * The group of instances according to hostPatterns defined in Zabbix rule file
```

**Comment by wu-sheng** [comment]
Besides https://github.com/apache/skywalking/pull/6176#discussion_r573495258, the PR looks good to me.

File: oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/protocol/ZabbixProtocolDecoder.java [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+        if (dataLength <= 0) {
+            throw new ZabbixErrorProtocolException("content could not be empty");
+        }
+
+        // Skip protocol extensions
+        byteBuf.skipBytes(4);
+
+        // Reading content
+        byte[] payload = new byte[dataLength];
+        byteBuf.readBytes(payload);
```
Comment (kezhenxu94) [inline oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/protocol/ZabbixProtocolDecoder.java:95]: This may cause memory leak (as well as line 67), please recheck the JavaDoc of `io.netty.handler.codec.ByteToMessageDecoder`.

```
 * Some methods such as {@link ByteBuf#readBytes(int)} will cause a memory leak if the returned buffer
 * is not released or added to the <tt>out</tt> {@link List}. Use derived buffers like {@link ByteBuf#readSlice(int)}
 * to avoid leaking memory.
```

File: oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/protocol/ZabbixProtocolDecoder.java [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+
+    /**
+     * Decode protocol to payload string
+     */
+    public String decodeToPayload(ChannelHandlerContext channelHandlerContext, ByteBuf byteBuf) throws InterruptedException, ZabbixErrorProtocolException {
+        int readable = byteBuf.readableBytes();
+        int baseIndex = byteBuf.readerIndex();
+        if (readable < HEADER_LEN) {
+            throw new ZabbixErrorProtocolException("header length is not enough");
+        }
```
Comment (kezhenxu94) [inline oap-server/server-receiver-plugin/skywalking-zabbix-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/zabbix/provider/protocol/ZabbixProtocolDecoder.java:63]: You cannot assume that there is enough bytes in the first round of decoding, you should reset the reader index and try another round until there is enough bytes

**Review by kezhenxu94 — DISMISSED** [review]

**Review by wu-sheng — APPROVED** [review]
LGTM. Thanks. Happy Chinese new year.

**Review by wu-sheng — DISMISSED** [review]

**Comment by wu-sheng** [comment]
UT fails, could you recheck?

**Review by wu-sheng — APPROVED** [review]
